src/rulesets/LinearAlgebra/structured.jl

```diff
@@ -47,6 +47,225 @@ end
 _symmetric_back(ΔΩ) = UpperTriangular(ΔΩ) + LowerTriangular(ΔΩ)' - Diagonal(ΔΩ)
 _symmetric_back(ΔΩ::Union{Diagonal,UpperTriangular}) = ΔΩ
 
+#####
+##### `Symmetric{<:Real}`/`Hermitian` eigendecomposition
+#####
+
+function frule((_, ΔA), ::typeof(eigen), A::LinearAlgebra.RealHermSymComplexHerm)
+    F = eigen(A)
+    ∂F = Thunk() do
+        λ, U = F
+        M = U' * ΔA * U
+        ∂λ = real(diag(M)) # if ΔA is Hermitian, so is U′ΔAU, so its diagonal is real
+        # K is skew-hermitian with zero diag
+        K = M ./ _nonzero.(λ' .- λ)
+        _setdiag!(K, Zero())
+        ∂U = U * K
+        return Composite{typeof(F)}(values = ∂λ, vectors = ∂U)
+    end
+    return F, ∂F
+end
+
+function rrule(::typeof(eigen), A::LinearAlgebra.RealHermSymComplexHerm)
+    F = eigen(A)
+    function eigen_pullback(ΔF)
+        ∂A = Thunk() do
+            λ, U = F
+            ∂λ, ∂U = ΔF.values, ΔF.vectors
+            # K is skew-hermitian
+            K = U' * ∂U
+            # unstable for degenerate matrices
+            U′∂AU = K ./ (_nonzero.(λ' .- λ))
+            setdiag!(U'∂AU, ∂λ)
+            return _symherm(A, U * U′∂AU * U')
+        end
+        return NO_FIELDS, ∂A
+    end
+    return F, eigen_pullback
+end
+
+function frule((_, ΔA), ::typeof(eigvals), A::LinearAlgebra.RealHermSymComplexHerm)
+    λ, U = eigen(A)
+    return λ, @thunk real(diag(U' * ΔA * U))
+end
+
+function rrule(::typeof(eigvals), A::LinearAlgebra.RealHermSymComplexHerm)
+    F, back = rrule(eigen, A)
+    λ, U = F
+    function eigvals_pullback(Δλ)
+        ∂A = Thunk() do
+            ∂F = Composite{typeof(F)}(values = Δλ)
+            return unthunk(back(∂F)[2])
+        end
+        return NO_FIELDS, ∂A
+    end
+    return λ, eigvals_pullback
+end
+
+_nonzero(x) = abs(x) > eps(eltype(x)) ? x : eps(eltype(x)) * sign(x)
```

Comment:
can you add a comment about this function? right now i'm not sure if it has the expected behaviour or not, 
since it won't make `0.0` non-zero (because `sign(0.0) === 0.0`)